### ISSUE TITLE
23502: null value in column "Id" violates not-null constraint

### ISSUE THREAD

**Issue by kuhnboy** [neisbut/Npgsql.Bulk#39]
Hello,
I'm getting an issue with an Insert operation against a semi-basic schema (23502: null value in column "Id" violates not-null constraint).

Here is my data structure:

```cs
public class Event
	{
		[Key]
		public Guid Id { get; set; }

		public Geometry Location { get; set; }

		public JsonElement ExtensionData { get; set; }
	}
```

**Comment by kuhnboy** [comment]
Insertion Code:

```cs
evts.Add(new Event { Id = Guid.NewGuid(), ExtensionData = doc.RootElement, Location = geom });

```

Every 500 or so events I will call Insert on the list of events.

**Comment by neisbut** [comment]
Hi @kuhnboy, is it possible to show the mapping code? That would help to undestand the issue. Also, what is the `evts.Add` ? How `NpgsqlBulkUploader` is called?

**Comment by raime910** [comment]
evts.Add(...) is what builds the IEnumerable<TEntity> I assume. I get the same thing here. I checked my entity made sure that the Id has a Guid value and still get this error message,

**Comment by neisbut** [comment]
Guys, @kuhnboy, @raime910, what EF and framework version you are using?

**Comment by raime910** [comment]
.net core 2.2 / ef 2.2

**Comment by weweilep** [comment]
I was able to reproduce this with a simple table with a Guid key and DateTimeOffset column.  There were different errors based on if the Guid was "generated" or "identity".  So it seems Guid is broken when it is a key in general.

**Comment by sadedil** [comment]
Same problem here.

We have a .net core / ef core 2.2 project and our models are using Guid PK's. Even if I set this values with `Guid.NewGuid()` before call `InsertAsync` i get this error (null value in column "Id" violates not-null constraint)

**Comment by brageej** [comment]
Facing the same issue with string keys.

**Comment by fahrigoktuna** [comment]
Hi All,
make sure that your EF entity Id property applies   [DatabaseGenerated(DatabaseGeneratedOption.None)] attribute. This fill wix your problems, issue can be closed.

**Comment by neisbut** [comment]
Hi guys, sorry for such a long silence period (but there was a reason for that). Thanks for all your feedback! Recently I finished a pretty big refactoring which should help with this issue. Now it is published as a 0.8-beta version. Please try when possible. Thanks!

**Comment by weweilep** [comment]
Using 0.8-beta using table

```
CREATE TABLE public."Entry"
(
    "Id" uuid NOT NULL,
    "Value" text COLLATE pg_catalog."default",
    CONSTRAINT "Entry_pkey" PRIMARY KEY ("Id")
)
```

Trying to insert the following entity:
```
using System;
using System.ComponentModel.DataAnnotations;

namespace bulktest
{
    public class Entry
    {
        [Key]
        public Guid Id { get; set; } = Guid.NewGuid();

        public string Value { get; set; } = "test";
    }
}

```

Results in:
`Sequence contains no matching element`
```

   at System.Linq.ThrowHelper.ThrowNoMatchException()
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source, Func`2 predicate)
   at Npgsql.Bulk.NpgsqlHelper.<>c__DisplayClass1_0.<GetMetadata>b__1(IProperty x)
   at System.Linq.Enumerable.SelectEnumerableIterator`2.ToList()
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Npgsql.Bulk.NpgsqlHelper.GetMetadata(DbContext context, Type type)
   at Npgsql.Bulk.NpgsqlBulkUploader.GetMappingInfo(Type type, String tableName)
   at Npgsql.Bulk.NpgsqlBulkUploader.CreateEntityInfo[T]()
   at Npgsql.Bulk.NpgsqlBulkUploader.<GetEntityInfo>b__44_0[T](Type x)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Npgsql.Bulk.NpgsqlBulkUploader.GetEntityInfo[T]()
   at Npgsql.Bulk.NpgsqlBulkUploader.Insert[T](IEnumerable`1 entities, InsertConflictAction onConflict)
   at Npgsql.Bulk.NpgsqlBulkUploader.Insert[T](IEnumerable`1 entities)
   at bulktest.Program.Main(String[] args) in C:\Projects\bulktest\Program.cs:line 23
```

**Comment by neisbut** [comment]
Hi @weweilep, thanks for testing this. I tried this test case on my side and I got an idea what could go wrong, but need your input. What table name was used in EF context? Was it exactly "Sample" or "Samples"? 
It seems the case when table name is specified incorrectly is handled a bit incorrecty and ugly error is thrown (will be fixed in next release). Same happened to me, but when I fixed table name it worked correctly.
So double check used table name please and try again.

**Comment by weweilep** [comment]
@neisbut It was "Sample" singular, and registered with the context as such.  However, I found my sample database was created sloppily (not using migrations) and the EF user account had access to the database, but not the table, so it appears that error might generally be related to failed permissions.  When I tried doing the insert with EF's AddRange instead of bulk, the error became obvious.

Once I granted proper access, the 0.8.0-beta is inserting correctly.

So it appears permission errors do not come across well.
